### PR TITLE
Extract row count and data size metadata from UnityCatalog

### PR DESCRIPTION
**Pull request by usefulalgorithm**
### 🤔 Why?

As titled.

### 🤓 What?

- Introduces new sql connector dependency
- Does `ANALYZE TABLE table_name COMPUTE STATISTICS` to get row count and data size

### 🧪 Tested?

- Did unit test.
- Tested end to end on personal dev env
截圖 2023-11-14 晚上10 59 25

### ☑️ Checks

- [x] My PR contains actual code changes, and I have updated the version number in `pyproject.toml`.

### CODE REVIEW

**Review by mars-lan — COMMENTED** [review]
Does this require additional permission?

**Comment by usefulalgorithm** [comment]
> Does this require additional permissions?

no, but I'm currently trying to figure out why it's hanging when I run it...

**Comment by mars-lan** [comment]
> > Does this require additional permissions?
> 
> no, but I'm currently trying to figure out why it's hanging when I run it...

It's most likely it's waiting for the warehouse to start. We'll need to update the README with additional information on that.

**Comment by usefulalgorithm** [comment]
# TODO

- [ ] separate statistics gathering logic to its own crawler (see snowflake profile crawler)
- [ ] warehouse id in config, if not specified then just use first warehouse

File: metaphor/unity_catalog/profile/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+This connector extracts dataset-level data profiles from Unity Catalog using the [Unity Catalog API](https://api-docs.databricks.com/rest/latest/unity-catalog-api-specification-2-1.html).
+
+## Setup
+
+Create an access token in the Databrick workspace > `User setting` > `Access tokens`.
+
+Data lineage should be anbled by default for all workspaces that reference Unity Catalog. You can manually enable it by following [these instructions](https://docs.databricks.com/data-governance/unity-catalog/enable-workspaces.html)
+
+> Refer to [this document](https://docs.databricks.com/data-governance/unity-catalog/data-lineage.html#limitations) to understand the limtations of Unity Catalog's data lineage.
```
Comment (mars-lan) [inline metaphor/unity_catalog/profile/README.md:11]: Should talk about the permissions needed.

File: tests/unity_catalog/test_extractor.py [hunk 2]
```diff
@@ -184,6 +186,25 @@ def mock_list_query_history(
     ]
     mock_create_api.return_value = mock_client
 
+    # Not a fan of all these mocks...
```
Comment (mars-lan) [inline tests/unity_catalog/test_extractor.py:189]: Remove?

File: metaphor/unity_catalog/profile/extractor.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            statistics=self._get_dataset_statistics(table_info=table_info),
+        )
+
+    def _get_dataset_statistics(self, table_info: TableInfo) -> DatasetStatistics:
+        dataset_statistics = DatasetStatistics()
+        if table_info.table_type is not TableType.VIEW:
+            with self._connection.cursor() as cursor:
+                # This can take a while
+                cursor.execute(
+                    f"ANALYZE TABLE {table_info.full_name} COMPUTE STATISTICS"
```
Comment (mars-lan) [inline metaphor/unity_catalog/profile/extractor.py:98]: Do we need to quote/escape the name if it contains special characters?

Comment (usefulalgorithm) [reply metaphor/unity_catalog/profile/extractor.py:98]: Yes, but the [documentation](https://docs.databricks.com/en/sql/language-manual/sql-ref-names.html) is extremely unclear:
> When referencing UC names in SQL, you must use backticks to escape names that contain special characters such as hyphens (-).
So what are the special characters other than `-`?

File: metaphor/unity_catalog/profile/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    directory: <output_directory>
+```
+
+See [Output Config](../common/docs/output.md) for more information on `output`.
+
+### Optional Configurations
+
+See [Filter Configurations](../common/docs/filter.md) for more information on the optional `filter` config.
+
+### Warehouse ID
```
Comment (mars-lan) [inline metaphor/unity_catalog/profile/README.md:33]:
```suggestion
#### Warehouse ID
```

File: metaphor/unity_catalog/profile/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+To run the connector with a specific warehouse, simply add its id in the configuration file:
```
Comment (mars-lan) [inline metaphor/unity_catalog/profile/README.md:35]:
```suggestion
To run the queries using a specific warehouse, simply add its ID in the configuration file:
```

File: metaphor/unity_catalog/profile/extractor.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                escaped_name = escape_special_characters(table_info.full_name)
+                # This can take a while
+                cursor.execute(f"ANALYZE TABLE {escaped_name} COMPUTE STATISTICS")
+                cursor.execute(f"DESCRIBE TABLE EXTENDED {escaped_name}")
+                rows = cursor.fetchall()
+                statistics = next(
+                    (row for row in rows if row.col_name == "Statistics"), None
+                )
+                if statistics:
+                    # `statistics.data_type` looks like "X bytes, Y rows"
```
Comment (mars-lan) [inline metaphor/unity_catalog/profile/extractor.py:106]: What a terriable way to return data in SQL 🙄

File: metaphor/unity_catalog/profile/utils.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+The following characters are not allowed:
+- `.`
+- ` `
+- `/`
+
+See https://docs.databricks.com/en/sql/language-manual/sql-ref-names.html for reference.
+"""
+
+
+def escape_special_characters(name: str) -> str:
```
Comment (mars-lan) [inline metaphor/unity_catalog/profile/utils.py:16]: Is there any downside of always quoting the name even if if contains no special characters?

**Review by mars-lan — COMMENTED** [review]
LGTM at a high level. Will let @elic-eon give the final ship it.

File: metaphor/unity_catalog/profile/extractor.py (hunk 6 above)
Comment (usefulalgorithm) [reply metaphor/unity_catalog/profile/extractor.py:106]: haha true

File: metaphor/unity_catalog/profile/utils.py (hunk 7 above)
Comment (usefulalgorithm) [reply metaphor/unity_catalog/profile/utils.py:16]: probably not, but the docs said to quote them only when they have special characters

Comment (usefulalgorithm) [reply metaphor/unity_catalog/profile/utils.py:16]: I guess let's stick with that for now?

**Review by elic-eon — APPROVED** [review]
LGTM

**Pull request merged**